zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -2396,12 +2395,17 @@ private void checkDataSizeLimit(String path, byte[] data) {
   }
 
   public void watchForData(final String path) {
-    watchForData(path, false);
+    watchForData(path, false, false);
   }
 
-  private boolean watchForData(final String path, boolean skipWatchingNonExistNode) {
+  public boolean watchForData(final String path, boolean skipWatchingNonExistNode, boolean persistListener) {
     try {
-      if (skipWatchingNonExistNode) {
+      if (persistListener) {
+        retryUntilConnected(() -> {
+          ((ZkConnection) getConnection()).getZookeeper().addWatch(path, AddWatchMode.PERSISTENT);
+          return true;
```

Comment:
neat: If we are anyways going to return "true" here means if execution enters into "persistListener" branch then it will never enter following branches right? in that case we can change it to "if" rather than "else if"?

Reply:
Not really, the `return true` is inside the lambda closure. We need to explicitly call return otherwise. Let's keep it as it is.